Fix of Code genertated from optional message field is mission optionality

Zetazzz opened the pull request:
Fix of https://github.com/cosmology-tech/telescope/issues/475

Use prototypes.fieldDefaultIsOptional: true to set all fields optional, except BingdingEntry fields for map<> fields.
If you want to set required fields according to [(gogoproto.nullable) = false] in proto files, prototypes.useOptionalNullable: true can be set while fieldDefaultIsOptional: true is set.

So this setting can solve the issue:
prototypes: {
    fieldDefaultIsOptional: true,
    useOptionalNullable: true, // false, if don't want to use  [(gogoproto.nullable) = false]
}

webmaster128 commented:
Thank you for looking into this @Zetazzz!

> Use prototypes.fieldDefaultIsOptional: true to set all fields optional

I think this is the wrong logic given how the protobuf spec works. All [Scalar Value Types](https://protobuf.dev/programming-guides/proto3/#scalar) should never be optional. For those, an unset field always falls back to the [default value](https://protobuf.dev/programming-guides/proto3/#default). So those should not be changed in here as the output was correct.

Only [Other Message Types](https://protobuf.dev/programming-guides/proto3/#other) in fields should default to optional. The `[(gogoproto.nullable) = false]` annotation is there to turn this from optional -> non-optional in cases where we know the field must always be set.

Zetazzz commented:
> Thank you for looking into this @Zetazzz!
> 
> > Use prototypes.fieldDefaultIsOptional: true to set all fields optional
> 
> I think this is the wrong logic given how the protobuf spec works. All [Scalar Value Types](https://protobuf.dev/programming-guides/proto3/#scalar) should never be optional. For those, an unset field always falls back to the [default value](https://protobuf.dev/programming-guides/proto3/#default). So those should not be changed in here as the output was correct.
> 
> Only [Other Message Types](https://protobuf.dev/programming-guides/proto3/#other) in fields should default to optional. The `[(gogoproto.nullable) = false]` annotation is there to turn this from optional -> non-optional in cases where we know the field must always be set. To quote the Default Values spec for those cases:
> 
> > For message fields, the field is not set. Its exact value is language-dependent. See the [generated code guide](https://protobuf.dev/reference/) for details.

Got it! Fixed!

I believe the unexpected results was because of Telescope consider the default value of (gogoproto.nullable) is false, so I added the option `optionalNullableDefaultTrue` to set it to true. Other logic's remain the same with before.

Proro:
```
message EvalRequest {
  // Bindings for the external variables.  The types SHOULD be compatible
  // with the type environment in [CheckRequest][google.api.expr.conformance.v1alpha1.CheckRequest], if checked.
  map<string, google.api.expr.v1alpha1.ExprValue> bindings = 1;

  map<string, google.api.expr.v1alpha1.IdRef> refs = 2;

  uint32 test_num = 3;
  string test_string = 4[(gogoproto.nullable) = false];
  bool test_bool = 5;

  // Used in v1beta1
  reserved 6, 7, 9 to 11;

  AccessConfig instantiate_permission = 8;

  oneof test_oneof {
    // [(gogoproto.nullable) = false] wouldn't work in this case
    string id = 12[(gogoproto.nullable) = false];
    string name = 13;
  }

  repeated string test_array = 14;
}
```

will generate TS:
```
export interface EvalRequest {
  /**
   * Bindings for the external variables.  The types SHOULD be compatible
   * with the type environment in [CheckRequest][google.api.expr.conformance.v1alpha1.CheckRequest], if checked.
   */
  bindings: {
    [key: string]: ExprValue;
  };
  refs: {
    [key: string]: IdRef;
  };
  testNum: number;
  testString: string;
  testBool: boolean;
  instantiatePermission?: AccessConfig;
  /** [(gogoproto.nullable) = false] wouldn't work in this case, since it's inside oneof */
  id?: string;
  name?: string;
  testArray: string[];
}
```

If you think this result's ok, I'll merge and publish. Thank you very much!

webmaster128 commented:
Thanks a lot! The diff looks good as far as I can tell.

I just wonder if we need the new option `optionalNullableDefaultTrue`. I think it should always be `true` because this is how protobuf works. Falling back to non-nullable for message fields is not very helpful.
